src/app/api/core/services/policies.service.ts

```diff
@@ -0,0 +1,30 @@
+import { UserAction, UserRole } from '@api/core/types/user'
+import { Resource } from '@api/core/types/api'
+import { BaseService } from './base.service'
+import APIError from '@api/core/exceptions/api'
+import httpStatus from 'http-status'
+
+/**
+ * Resource-level policies control service for Users (IU + Client)
+ */
+export class PoliciesService extends BaseService {
+  private defaultPolicies: Record<UserRole.Client, Record<Resource, UserAction[]>> = {
+    [UserRole.Client]: {
```

Comment:
Can we check policies based on the resources and check role later on? that we if we need to check policies for internal users, then we wont have to add keys.

And also just checking actions (read, create) is not scalable. 

I think a bit scalable solution is:
- Create policies on resource label
- Create a Policy file for each resource
- Create methods with the name of actions required, in the Policy file

Eg:

```
// policies.ts
export const defaultPolicies: Record<UserRole.Client, Record<Resource, UserAction[]>> = {
      [Resource.Tasks]: TaskPolicy,
      [Resource.WorkflowState]: WorkflowStatePolicy,
  }
```

```
// PolicyGate.ts
export class PolicyGate {

  // ..

  authorize(action: PolicyAction, resource: Resource) {
    const policy = new defaultPolicies[resource]();
    // call method based on the action here
  }
}
```

```
// TaskPolicy.ts
export class TaskPolicy {
  read(resource: Resource, user: User) {
    if (user.role === UserRole.IU) {
      return true;
    }
    
    if (user is in the matched company OR task assignee is the client) {
     return true;
    }
    
    return false;
  }
}
```

Let me know what you think? I think this looks a bit clean and scalable

Reply:
@sazanrjb Your recommendation definitely makes sense and is cleaner but it also adds a lot of boilerplate for each resource. I think we should be good with the current approach at least for now... we can think up a solution or refactor to this when this requirement comes (if it does).

My idea was we could create a `Permissions` table and query permissions with `db.permissions.findFirst({ where: { resource })` instead of using `defaultPolicies` to authorize later on. This way we don't need individual `xxx.policy.ts` files and repeat ourselves.